include/libtorrent/ip_notifier.hpp

```diff
@@ -0,0 +1,81 @@
+/*
+
+Copyright (c) 2016, Steven Siloti
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+* Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+* Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in
+the documentation and/or other materials provided with the distribution.
+* Neither the name of the author nor the names of its
+contributors may be used to endorse or promote products derived
+from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_IP_NOTIFIER_HPP_INCLUDED
+#define TORRENT_IP_NOTIFIER_HPP_INCLUDED
+
+#include <functional>
+#include <array>
+
+#include "libtorrent/error_code.hpp"
+#include "libtorrent/address.hpp"
+#include "libtorrent/io_service.hpp"
+
+#if defined TORRENT_BUILD_SIMULATOR
+#elif TORRENT_USE_NETLINK
+#include "libtorrent/netlink.hpp"
+#elif defined TORRENT_WINDOWS
+#include <boost/asio/windows/object_handle.hpp>
+#endif
+
+namespace libtorrent
+{
+	struct ip_change_notifier : boost::noncopyable
+	{
+		// cb will be invoked from within a thread calling ios.run()
+		// when a change is detected in the system's IP addresses
+		ip_change_notifier(io_service& ios, std::function<void()> cb);
+		~ip_change_notifier();
+
+		// start listening for changes
+		void start();
```

Comment:
essentially all of libtorrent uses this pattern of "subscribing" to events, as opposed to the underlying asio model where you can only ask for one-shot events, and then re-subscribe. This has subtle implications of lifetime management on the object. I think in general, asio's way is more composable. It specifically simplifies shutting down. In this case, since the io_service is tied to the session object, the session destructor is synchronized with the io_service running out of work, which makes it all good. But in the simulator case, there can be multiple sessions sharing an io_service, which means it's critical that all handlers have been called before we can destruct the session. With a subscription interface, calling stop, there's no way of knowing whether there's an outstanding handler to be called afterwards, so it's not safe to destruct the session (and the ip_change_notifier along with it). The way I've dealt with this problem in ``lsd``, ``natpmp`` and ``upnp`` is to make those objects be heap allocated, managed by a ``shared_ptr`` and make the handler keep them alive. So, even it the session_object has destructed, the object is still alive to support receiving the last callback.

The problem with this approach is that every level of io-object needs to be heal allocated to deal with clean shutdown.

I don't think it's that big of a deal in this specific case, since the session object already suffers from this problem from code that would take a lot more to fix. However, you may want to consider moving the callback object from being passed into the constructor to the ``start()`` function and have the ``stop()`` function actually destruct it. That lets the callback keep something alive with a reference count. It probably means you need a boost::optional<std::function<void()>> too though.

just a thought.